state/machine.go

```diff
@@ -4,325 +4,524 @@ import (
+		signedBlock := block.SignedBlock{}
+		if machine.validValue != nil {
+			signedBlock = *machine.validValue
+		} else {
+			signedBlock = machine.buildSignedBlock()
+		}
+
+		if signedBlock.Height != machine.currentHeight {
+			panic("unexpected block")
+		}
+
+		propose := block.Propose{
+			Block:      signedBlock,
+			Round:      round,
+			ValidRound: machine.validRound,
+			LastCommit: machine.lastCommit,
+		}
 
-func (machine *machine) InsertPrecommit(precommit block.SignedPreCommit) {
-	machine.commitBuilder.Insert(precommit)
+		signedPropose, err := propose.Sign(machine.signer)
+		if err != nil {
+			panic(err)
+		}
+
+		return Propose{
+			SignedPropose: signedPropose,
+			Commit:        committed,
+		}
+	}
+
+	if len(committed.Signatures) > 0 {
+		return Commit{
+			Commit: *commit,
+		}
+	}
+	return nil
 }
 
 func (machine *machine) SyncCommit(commit block.Commit) {
-	if commit.Polka.Height > machine.height {
-		machine.state = WaitingForPropose{}
-		machine.height = commit.Polka.Height + 1
-		machine.round = 0
-		machine.lockedBlock = nil
-		machine.lockedRound = nil
+	if commit.Polka.Height >= machine.currentHeight {
+		machine.currentState = WaitingForPropose{}
+		machine.currentHeight = commit.Polka.Height + 1
+		machine.currentRound = 0
+		machine.lockedValue = nil
+		machine.lockedRound = -1
+		machine.validValue = nil
+		machine.validRound = -1
+		machine.lastCommit = &commit
+		machine.bufferedMessages = map[block.Round]map[sig.Signatory]struct{}{}
+		machine.ticksAtProposeState = 0
+		machine.ticksAtPrevoteState = -1
+		machine.ticksAtPrecommitState = -1
+		machine.drop()
 	}
 }
 
 func (machine *machine) Transition(transition Transition) Action {
 	// Check pre-conditions
-	if machine.lockedRound == nil {
-		if machine.lockedBlock != nil {
+	if machine.lockedRound < 0 {
+		if machine.lockedValue != nil {
 			panic("expected locked block to be nil")
 		}
 	}
-	if machine.lockedRound != nil {
-		if machine.lockedBlock == nil {
+	if machine.lockedRound >= 0 {
+		if machine.lockedValue == nil {
 			panic("expected locked round to be nil")
 		}
 	}
 
-	switch machine.state.(type) {
+	if machine.validRound < 0 {
+		if machine.validValue != nil {
+			panic("expected valid block to be nil")
+		}
+	}
+	if machine.validRound >= 0 {
+		if machine.validValue == nil {
+			panic("expected valid round to be nil")
+		}
+	}
+
+	if transition.Round() > machine.currentRound {
+		if _, ok := machine.bufferedMessages[transition.Round()]; !ok {
+			machine.bufferedMessages[transition.Round()] = map[sig.Signatory]struct{}{}
+		}
+		if _, ok := machine.bufferedMessages[transition.Round()][transition.Signer()]; !ok {
+			machine.bufferedMessages[transition.Round()][transition.Signer()] = struct{}{}
+		}
+	}
+
+	higherRound := machine.checkForHigherRounds()
+	if higherRound != nil && *higherRound > machine.currentRound {
+		switch transition := transition.(type) {
+		case PreVoted:
+			_ = machine.polkaBuilder.Insert(transition.SignedPreVote)
```

Comment:
I prefer having the `_ =` because it explicitly tells the reader: “there’s a value being returned by I’m ignoring”. For example, even in this review, by seeing this there is the (rightful) question: “why is this value ignored?”

@divyakoshy why is this value being ignored? If there’s good justification, please add a short explicit comment above this assignment.

Reply:
I have added justification for all ignored return values.